### PR TITLE
Remove observers for value and errors

### PR DESCRIPTION
**Pull request by poteto**
This PR:
- [x] Removes the `fieldObserver` observer that was watching changes in
  `value`. This has been replaced with a CP that has both a getter and
  setter.
- [x] Removes the `setErrorsObserver` observer. This has been replaced with
  a CP that is added at runtime when the component's `didInitAttrs` hook
  is called.
- [x] Adds the ability to watch for errors on the Model (via DS.Errors).
- [x] Makes `ya-form` a dependency.
- [x] Don't use `setProperties`
- [x] Adds a basic acceptance test

Will work on this more tomorrow!

Note: It is easier to read the diff in [split mode](https://github.com/danmcclain/ya-input/pull/3/files?diff=split)

### CODE REVIEW

**Comment by brzpegasus** [comment]
:clap: for removing observers!

**Comment by danmcclain** [comment]
:+1: Thanks!

File: addon/components/ya-input/component.js [hunk 1]
```diff
@@ -3,89 +3,86 @@ import layout from '../../templates/ya-input';
...
-  layout: layout,
+  layout,
+  type: 'text',
   classNameBindings: ['validClass', 'type'],
-
-  field: '',
-  canShowErrors: false,
+  focusedOnce: false,
   hasLostFocus: false,
+  initialFocus: false,
```
Comment (michaeldupuisjr) [inline addon/components/ya-input/component.js]: I think booleans are a lot more expressive when prepended by `is` or `has`.

For example, I'm not sure by the name of this property, what it's keeping track of. Is it keeping track of something having been focused on at a point in the past? If so, `hasReceivedFocus` would be more meaningful IMO.

File: addon/components/ya-input/component.js [hunk 2]
```diff
@@ -3,89 +3,86 @@ import layout from '../../templates/ya-input';
...
-      const shouldShowValidationErrors = get(this, 'form.model.shouldShowValidationErrors');
+      const shouldShowValidationErrors = get(this, 'shouldShowValidationErrors');
 
-      if (errorsCount === 0) {
-        set(this, 'canShowErrors', false);
-      } else if (errorsCount > 0 && (hasLostFocus || shouldShowValidationErrors)) {
-        set(this, 'canShowErrors', true);
-      }
-    });
+      return !!(errorsCount > 0 && (hasLostFocus || shouldShowValidationErrors));
```
Comment (michaeldupuisjr) [inline addon/components/ya-input/component.js:72]: I know this `shouldShowValidationErrors` already existed, but I find it pretty confusing to have a function named `canShowErrors` and another one called `shouldShowValidationErrors` – nevermind having one called within the other. By name, you would think these functions served the same purpose.

One of these functions should be renamed to more accurately describe what it's doing.

File: addon/components/ya-input/component.js [hunk 3]
```diff
@@ -97,5 +94,67 @@ export default Component.extend({
...
+  _getValidClass(modelErrors, fieldName) {
+    if (get(this, 'focusedOnce') || get(this, 'shouldShowValidationErrors')) {
+      if (this._hasDSError(modelErrors, fieldName)) {
+        return 'is-invalid';
+      }
+
+      return get(this, 'canShowErrors') ? 'is-invalid' : 'is-valid';
+    }
+
+    return;
```
Comment (michaeldupuisjr) [inline addon/components/ya-input/component.js]: Is this `return` necessary?

File: tests/acceptance/main-test.js [hunk 4]
```diff
@@ -0,0 +1,42 @@
...
+    run(this.application, 'destroy');
+  }
+});
+
+test('it displays validation errors', (assert) => {
+  assert.expect(4);
+
+  return new MainRoute(assert, { routeName: '/' })
+    .assertVisitUrl()
+    .firstName('Milton')
```
Comment (michaeldupuisjr) [inline tests/acceptance/main-test.js]: I think this function needs a verb. For example, are we filling in here or asserting?

Ok – I know that we're filling in by looking at the PageObject helper, but I find this naming very un-intuitive. I would prefer `fillInName('Milton')` or `fillInByLabel('name', 'Milton')`. Maybe I'm misunderstanding the PageObject library in general, but it seems that it is trying to do too much by trying to handle finders, interactions, and assertions. I would think that the library should simply take elements in the DOM and make them objects for our tests. You could persuade me to implement some `find` methods in there as well, but I definitely think assertions and interactions should be extracted.

I also think this test should ensure that the `firstName()` actually filled in the field. The `assertIsValid()` later is of little use if the field itself is not in the expected state.

**Comment by poteto** [comment]
:eyes: please @danmcclain

File: addon/components/ya-input/component.js [hunk 5]
```diff
@@ -3,99 +3,157 @@ import layout from '../../templates/ya-input';
...
 
-  errorText: oneWay('errors.firstObject'),
+  model: oneWay('form.model'),
+  modelName: oneWay('model.constructor.modelName'),
+  shouldShowValidationErrors: oneWay('model.shouldShowValidationErrors'),
 
-  fieldObserver: observer('value', function() {
-    const props  = getProperties(this, 'form.model', 'field-name');
-    set(props['form.model'], props['field-name'], get(this, 'value'));
+  errorText: computed('errors.[]', {
```
Comment (danmcclain) [inline addon/components/ya-input/component.js]: I prefer to show one error at a time

Comment (poteto) [inline addon/components/ya-input/component.js]: :+1:

**Pull request merged**